storage: add AddSSTable command to link a preconstructed file into RocksDB

danhhz opened the pull request:
This is the implementation of https://github.com/cockroachdb/cockroach/pull/15895.

A command is added that takes the contents of an sstable as constructed
by a RocksDB SstFileWriter (or our RocksDBSstFileWriter wrapper).

The propser verifies that the keys in the sstable are within the range
specified by the request header, verifies the key-value checksums, and
computes the new MVCCStats.

On each replica, as a side effect of application, the file is linked
directly into RocksDB using its IngestExternalFile call.

---

A new cluster setting 'kv.import.addsstable.enabled' is added to
optionally use the new AddSSTable command to implement Import (and thus
Restore). This should lead to reduced disk usage and eventually
dramatically increased throughput.

The setting defaults to false for now while we finalize the
implementation of AddSSTable (we're looking to avoid writing the file
contents in the normal RocksDB raft log to save even more disk usage)
and while we run it through production testing.

A test is added with runs the usual backupAndRestore test but with the
setting turned on.

This PR also introduces a hook to disable the use of RocksDB range
deletiongs. This works around a limitation in RocksDB that currently
does lot allow ingesting a file while a snapshot is outstanding. There
is an upcoming fix in RocksDB which will let us remove this workaround.

---

This is still waiting on a fix from RocksDB and the AddSSTable command wants some tests (it is tested a bit in TestBackupRestoreAddSSTable), but I wanted to get it out for early review. Note that this still stores the sstable data in RocksDB on behalf of the raft log, @tschottdorf has an upcoming rfc to address that

danhhz committed:
storage: add AddSSTable command to link a preconstructed file into RocksDB

This is the implementation of https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/add_sstable.md.

A command is added that takes the contents of an sstable as constructed
by a RocksDB SstFileWriter (or our RocksDBSstFileWriter wrapper).

The propser verifies that the keys in the sstable are within the range
specified by the request header, verifies the key-value checksums, and
computes the new MVCCStats.

On each replica, as a side effect of application, the file is linked
directly into RocksDB using its IngestExternalFile call.

danhhz committed:
storageccl: optionally use AddSSTable to implement Ingest

A new cluster setting 'kv.import.addsstable.enabled' is added to
optionally use the new AddSSTable command to implement Import (and thus
Restore). This should lead to reduced disk usage and eventually
dramatically increased throughput.

The setting defaults to false for now while we finalize the
implementation of AddSSTable (we're looking to avoid writing the file
contents in the normal RocksDB raft log to save even more disk usage)
and while we run it through production testing.

A test is added with runs the usual backupAndRestore test but with the
setting turned on.

This PR also introduces a hook to disable the use of RocksDB range
deletions. This works around a limitation in RocksDB that currently
does lot allow ingesting a file while a snapshot is outstanding. There
is an upcoming fix in RocksDB which will let us remove this workaround.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16158)
<!-- Reviewable:end -->

danhhz commented:
Review status: 0 of 21 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/replica_proposal.go, line 500 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl11Qr2n3V2426YCSI1:-Kl11Qr2n3V2426YCSI2:b-t9ai1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L500)):*
> ```Go
> 	var path string
> 	var move bool
> 	if inmem, ok := r.store.engine.(engine.InMem); ok {
> ```

This should probably be more of a hardlink sniff. Unfortunately the in-memory env's LinkFile call currently succeeds but is broken

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Commit message typo: deletiongs

---

Reviewed 17 of 18 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/storageccl/add_sstable.go, line 91 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5fE8VEJa-oRVjMcLb:-Kl5fE8VEJa-oRVjMcLc:b-g9s6o9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/ccl/storageccl/add_sstable.go#L91)):*
> ```Go
> 			// TODO(dan): Add a new field in roachpb.Error, so the client can
> 			// catch this and retry. It can happen if creating the file to
> 			// ingest races with a range split.
> ```

Races with splits should be detected before this point (since this is a range request we validate that the declared start and end keys are in bounds for the range). This is a guard against the accidental or malicious construction of an sstable that doesn't match the command's declared range.

---

*[pkg/ccl/storageccl/import.go, line 198 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5hLoCpEkZTdBdtv2x:-Kl5hLoCpEkZTdBdtv2y:b7sap3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/ccl/storageccl/import.go#L198)):*
> ```Go
> 	tempPrefix := filepath.Join(
> 		cArgs.EvalCtx.GetTempPrefix(),
> 		strconv.Itoa(int(parser.GenerateUniqueInt(cArgs.EvalCtx.NodeID()))),
> ```

You can use `ioutil.TempDir` here.

---

*[pkg/roachpb/api.go, line 939 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5cTXd6X0n968_N3we:-Kl5cTXeam9NBSGSSBhX:bqfyiwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/roachpb/api.go#L939)):*
> ```Go
> func (*ImportRequest) flags() int                   { return isAdmin | isAlone }
> func (*AdminScatterRequest) flags() int             { return isAdmin | isAlone | isRange }
> func (*AddSSTableRequest) flags() int               { return isWrite | isRange }
> ```

Do we need some sort of "unsplittable" flag so that DistSender won't try to duplicate this request if it sees there's been a split?

---

*[pkg/storage/replica_proposal.go, line 486 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5_wuvjPlSjPEQUHPT:-Kl5_wuvjPlSjPEQUHPU:b-4el4yh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L486)):*
> ```Go
> }
> 
> var addMu syncutil.Mutex
> ```

Looks unused.

---

*[pkg/storage/replica_proposal.go, line 501 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5axYSA7XRX_nQbomB:-Kl5axYSA7XRX_nQbomC:b6er8lm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L501)):*
> ```Go
> 	var move bool
> 	if inmem, ok := r.store.engine.(engine.InMem); ok {
> 		path = fmt.Sprintf("%x", checksum)
> ```

Similar to the other PR, should the path be a counter instead of a checksum?

I think you could get in trouble if the same sstable file were proposed to two different ranges on the same store. (This would never be a legitimate thing to do but could have nasty consequences)

---

*[pkg/storage/replica_proposal.go, line 507 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5c-4kDNQ4CdAfmjbZ:-Kl5c-4kDNQ4CdAfmjb_:b-ncimdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L507)):*
> ```Go
> 		}
> 	} else {
> 		path = filepath.Join(r.store.engine.GetTempDir(), fmt.Sprintf("addsstable-%x", checksum))
> ```

I think this could use `ioutil.TempFile(GetTempDir(), "addsstable")` to let go pick a guaranteed-available filename.

---

*[pkg/storage/replica_proposal.go, line 509 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5c95UopfXvDQdK49G:-Kl5c95UopfXvDQdK49H:b-qp8a54) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L509)):*
> ```Go
> 		path = filepath.Join(r.store.engine.GetTempDir(), fmt.Sprintf("addsstable-%x", checksum))
> 		move = true
> 		if err := ioutil.WriteFile(path, data, 0644); err != nil {
> ```

Any reason for the permissions to be broader than 0600?

---

*[pkg/storage/replica_proposal.go, line 739 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5eVxv9J_qZU2v8GB9:-Kl5eVxv9J_qZU2v8GBA:bjn9pjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L739)):*
> ```Go
> 
> 	if rResult.AddSSTable != nil {
> 		r.addSSTablePostApply(ctx, rResult.AddSSTable)
> ```

ReplicatedEvalResult is for *in-memory* side effects. If there was a crash before this line, your database would be corrupted because the command would be marked as applied but the new sstable wouldn't be there. We need to figure out how to make this atomic with the commit of the batch including the new AppliedIndex. I'm not sure if rocksdb has the right interfaces to make this possible.

---

*[pkg/storage/storagebase/proposer_kv.proto, line 87 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5_OPOM7pyJPRrMU4q:-Kl5_OPPXEwJhUH_CunD:b-ppp0im) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/storagebase/proposer_kv.proto#L87)):*
> ```protobuf
>   optional ChangeReplicas change_replicas = 12;
>   optional int64 raft_log_delta = 13;
>   optional bytes add_sstable = 16 [(gogoproto.nullable) = false, (gogoproto.customname) = "AddSSTable"];
> ```

I'd wrap this in a message just in case we have some metadata to send along with it in the future. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-:-Kl5Z1O4JIXXAfG-n7eb:b-920e73)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on pkg/roachpb/api.go:
```diff
@@ -927,6 +936,7 @@ func (*WriteBatchRequest) flags() int               { return isWrite | isRange }
 func (*ExportRequest) flags() int                   { return isRead | isRange | updatesTSCache }
 func (*ImportRequest) flags() int                   { return isAdmin | isAlone }
 func (*AdminScatterRequest) flags() int             { return isAdmin | isAlone | isRange }
+func (*AddSSTableRequest) flags() int               { return isWrite | isRange }
```
`isAlone`

tschottdorf commented on pkg/storage/replica_proposal.go:
```diff
@@ -470,6 +483,39 @@ func (r *Replica) leasePostApply(
+func (r *Replica) addSSTablePostApply(ctx context.Context, data []byte) {
+	var checksum []byte
+	{
+		hash := sha512.New()
```
I think we settled on CRC32.

danhhz commented:
Thanks for the reviews! Two more small BEFORE MERGE bits to fix up but this should be ready for another look.

I'd like to get this merged as an experimental feature while we investigate the resolution to the commit index + ingest atomicity. Having it in will make it easier to collaborate with @tschottdorf on the implementation of https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/raft_sstable_sideloading.md (as well as make it easier to start running some production load tests)

---

Review status: 6 of 23 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/storageccl/add_sstable.go, line 91 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5fE8VEJa-oRVjMcLb:-KltuKDRFQX9mnMZwzVU:bfczpfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/ccl/storageccl/add_sstable.go#L91)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Races with splits should be detected before this point (since this is a range request we validate that the declared start and end keys are in bounds for the range). This is a guard against the accidental or malicious construction of an sstable that doesn't match the command's declared range.
</blockquote></details>

Yeah, that's what I was intending this to say. Reworked it a bit

---

*[pkg/ccl/storageccl/import.go, line 198 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5hLoCpEkZTdBdtv2x:-KlVNxLiYToDXVJWGxhO:bypw42c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/ccl/storageccl/import.go#L198)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

You can use `ioutil.TempDir` here.
</blockquote></details>

Deleted all of this once I rebased on top of the in-mem sst writer

---

*[pkg/roachpb/api.go, line 939 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl11HeFuyiQzUTxf5wR-r2-939:-Klu8Fzfmto0_e3ms1uW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/roachpb/api.go#L939)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`isAlone`
</blockquote></details>

Done.

---

*[pkg/roachpb/api.go, line 939 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5cTXd6X0n968_N3we:-KltuahI_Xro_MliPa7h:b-rj6a69) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/roachpb/api.go#L939)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Do we need some sort of "unsplittable" flag so that DistSender won't try to duplicate this request if it sees there's been a split?
</blockquote></details>

Yeah, this could be nice and would also be applicable to WriteBatch. I filed an issue: https://github.com/cockroachdb/cockroach/issues/16338

---

*[pkg/storage/replica_proposal.go, line 486 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5_wuvjPlSjPEQUHPT:-KlVOErZrbxsYVKOGe8l:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L486)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Looks unused.
</blockquote></details>

Done

---

*[pkg/storage/replica_proposal.go, line 491 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl11HeFuyiQzUTxf5wb-r2-491:-Klu8IadkpiJEp3dZdpf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L491)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I think we settled on CRC32.
</blockquote></details>

Done.

---

*[pkg/storage/replica_proposal.go, line 501 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5axYSA7XRX_nQbomB:-KlVOIEI5pmpj_Lg2L3F:b6n5dab) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L501)):*
> I think you could get in trouble if the same sstable file were proposed to two different ranges on the same store.

sstables are required to have at least one entry and the proposer checks that the included keys are all within the right range, so I don't see how that could happen. that said, I see your point.

I left this as a TODO since toby's going to rewrite it anyway

---

*[pkg/storage/replica_proposal.go, line 507 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5c-4kDNQ4CdAfmjbZ:-Klu0hf7F8GORNLNlZF2:bb7goch) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L507)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think this could use `ioutil.TempFile(GetTempDir(), "addsstable")` to let go pick a guaranteed-available filename.
</blockquote></details>

Ditto TODO

---

*[pkg/storage/replica_proposal.go, line 509 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5c95UopfXvDQdK49G:-KlVOn48v_KaKUsLbXSn:bb710yj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L509)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Any reason for the permissions to be broader than 0600?
</blockquote></details>

Nope

---

*[pkg/storage/replica_proposal.go, line 739 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5eVxv9J_qZU2v8GB9:-Klu0j-I2pmJ-pc8FkG1:b-nchc7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L739)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

ReplicatedEvalResult is for *in-memory* side effects. If there was a crash before this line, your database would be corrupted because the command would be marked as applied but the new sstable wouldn't be there. We need to figure out how to make this atomic with the commit of the batch including the new AppliedIndex. I'm not sure if rocksdb has the right interfaces to make this possible.
</blockquote></details>

I'd like to leave this as followup work since there's some investigation between the two approaches. Having all the rest of this checked in will keep Toby and I from duplicating a lot of work

---

*[pkg/storage/storagebase/proposer_kv.proto, line 87 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl5_OPOM7pyJPRrMU4q:-Klu970GkAIHm0iP8Z52:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/storagebase/proposer_kv.proto#L87)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'd wrap this in a message just in case we have some metadata to send along with it in the future. 
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-:-KluAcEa5t2-HsBh6kRD:bn2ughr)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Cleaned up the last two bits

---

Review status: 6 of 23 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-:-KluNeEeHmIktSNih8dB:bsrl8gd)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 13 of 17 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-:-KluRDQPwf3qvb0CIcvU:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Didn't give this the most thorough review, but looked good generally. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 6 of 18 files at r1, 17 of 17 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_proposal.go, line 500 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl11Qr2n3V2426YCSI1:-KluhA0foYzJ4qzAeRib:b-vjmadi) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L500)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

This should probably be more of a hardlink sniff. Unfortunately the in-memory env's LinkFile call currently succeeds but is broken
</blockquote></details>

You mean `IngestExternalFile` below returns no error but the data isn't ingested?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-:-KlujMeEJOTOq3zu3d2e:b-ml9d35)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on pkg/storage/replica_proposal.go:
```diff
@@ -470,6 +482,39 @@ func (r *Replica) leasePostApply(
+func (r *Replica) addSSTablePostApply(ctx context.Context, data []byte) {
+	var checksum []byte
+	{
+		hash := crc32.New(crc32.MakeTable(crc32.Castagnoli))
```
I somehow imagined we'd compute the checksum along with the SSTable and send with the request. That way, we catch more corruption and besides, it's more opportune to do it where we create the SSTable in the first place, no?

tschottdorf commented:
BTW, I'm rebased on top of this now. So, you know, no sudden movements, please 😃

tschottdorf commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/add_sstable.go, line 51 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-KlxXkYTsZ1DqSWEZuUV:-KlxXkYTsZ1DqSWEZuUW:beok8nt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea3b2c499d30beab8370c7dc610a61f97b6f9545/pkg/ccl/storageccl/add_sstable.go#L51)):*
> ```Go
> 	// MVCCStats.
> 	mvccStartKey, mvccEndKey := engine.MVCCKey{Key: args.Key}, engine.MVCCKey{Key: args.EndKey}
> 	stats, err := verifySSTable(args.Data, mvccStartKey, mvccEndKey, h.Timestamp.WallTime)
> ```

Similar to the hash, wouldn't it be more direct to compute the stats and send them along with the request as well?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158)*
<!-- Sent from Reviewable.io -->

danhhz commented:
TFTRs!

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/add_sstable.go, line 51 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-KlxXkYTsZ1DqSWEZuUV:-KlxlAL_9-1vkH2pfDOM:b-kcre5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea3b2c499d30beab8370c7dc610a61f97b6f9545/pkg/ccl/storageccl/add_sstable.go#L51)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Similar to the hash, wouldn't it be more direct to compute the stats and send them along with the request as well?
</blockquote></details>

Same question as below about what you mean by "request". Let's figure this out in a followup.

---

*[pkg/storage/replica_proposal.go, line 500 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl11Qr2n3V2426YCSI1:-Klxj-YShrBDpbrCBhCG:b-98ofdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f93686ae4d7e5371caab4c95345210cb787a5318/pkg/storage/replica_proposal.go#L500)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

You mean `IngestExternalFile` below returns no error but the data isn't ingested?
</blockquote></details>

IngestExternalFile internally calls LinkFile on the in-mem env, which returns without an error but doesn't actually work. Then later in the IngestExternalFile impl, it tries to use the file, can't find it, and returns an error about a missing file (at the path it expects it in the rocksdb dir, not the original path). It's quite confusing to a caller of IngestExternalFile

---

*[pkg/storage/replica_proposal.go, line 488 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl11HeFuyiQzUTxf5wb-r4-488:-Klxk45JN4C3aCOKn9xO:b-y9avyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea3b2c499d30beab8370c7dc610a61f97b6f9545/pkg/storage/replica_proposal.go#L488)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I somehow imagined we'd compute the checksum along with the SSTable and send with the request. That way, we catch more corruption and besides, it's more opportune to do it where we create the SSTable in the first place, no?
</blockquote></details>

I was 50/50 on this. As far as I know, we don't do this in any other requests, but this one is expected to be much bigger so maybe it makes sense for it to be an exception. Unfortunately, WriteBatch already doesn't send a checksum, but arguably should if we add one to the AddSSTable request. This would be a bit annoying to add in a backward-compatible way, though it could be done (ignore empty checksum, but validate if it's there).

Sorry if this is a precisely defined term and I just don't have the context... what do you mean by request? The caller of (client.DB).AddSSTable or the evaluator?

Ultimately, I'd prefer to merge this as-is, figure out how this all interacts with your raft sideloading work, and address it in a followup. I'm happy to do the work of that followup.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-:-Klxls_o_11QP18PuUPE:bz5mair)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/add_sstable.go, line 51 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-KlxXkYTsZ1DqSWEZuUV:-Kly55aNFL4XfXV7qmzk:b-m1q3cb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea3b2c499d30beab8370c7dc610a61f97b6f9545/pkg/ccl/storageccl/add_sstable.go#L51)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Same question as below about what you mean by "request". Let's figure this out in a followup.
</blockquote></details>

`AddSSTableRequest`.

---

*[pkg/storage/replica_proposal.go, line 488 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl11HeFuyiQzUTxf5wb-r4-488:-KlyCUPdXko45XJUsXjP:b-z5ky7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea3b2c499d30beab8370c7dc610a61f97b6f9545/pkg/storage/replica_proposal.go#L488)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I was 50/50 on this. As far as I know, we don't do this in any other requests, but this one is expected to be much bigger so maybe it makes sense for it to be an exception. Unfortunately, WriteBatch already doesn't send a checksum, but arguably should if we add one to the AddSSTable request. This would be a bit annoying to add in a backward-compatible way, though it could be done (ignore empty checksum, but validate if it's there).

Sorry if this is a precisely defined term and I just don't have the context... what do you mean by request? The caller of (client.DB).AddSSTable or the evaluator?

Ultimately, I'd prefer to merge this as-is, figure out how this all interacts with your raft sideloading work, and address it in a followup. I'm happy to do the work of that followup.


</blockquote></details>

Yeah, no problem. Let's see how this looks when things come together.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_proposal.go, line 488 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158#-Kl11HeFuyiQzUTxf5wb-r4-488:-KlyhV6u2eV7t8XOnB6I:bgihybr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea3b2c499d30beab8370c7dc610a61f97b6f9545/pkg/storage/replica_proposal.go#L488)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yeah, no problem. Let's see how this looks when things come together.
</blockquote></details>

We already do end-to-end checksums (the `internal/client` package calls in to `roachpb.NewPut` which calls `roachpb.Value.InitChecksum`), so there's some precedent for adding a checksum field to AddSSTableRequest. But the most important aspect of checksums is just the way they're being used here. Data generally flows from the TLS-protected network connection to the checksummed WAL to the checksummed sstables. The interval between writing a sideloaded sstable to disk and ingesting it into rocksdb is a gap during which we are solely responsible for the integrity of the data, which is why we want to checksum things here (although the sstable does have internal checksums that rocksdb will validate on import)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16158)*
<!-- Sent from Reviewable.io -->
